### ISSUE TITLE
heatmap polar

### ISSUE THREAD

**Issue by CarolinGao12** [jheinen/GR.jl#103]
Is there someway that draw  heatmap to be polar axes  ?

**Comment by jheinen** [comment]
Sorry, that's not yet supported. I put it on my TODO list ...

**Comment by juliohm** [comment]
@jheinen please bump this feature in your TODO list :heart: 

I would like to use GR to generate the plot below, it is sometimes called the variogram plane plot:

![varplane](https://user-images.githubusercontent.com/3345261/49938756-f9ecf780-fec1-11e8-9c52-64afca9997ef.png)

I am able to generate it with the PyPlot backend, but GR is my favorite.

**Comment by jheinen** [comment]
Could you share the script + required date anywhere?

**Comment by juliohm** [comment]
This [notebook](http://nbviewer.jupyter.org/github/juliohm/GeoStatsTutorials/blob/master/notebooks/DirectionalVariograms.ipynb) should be reproducible. I am still working on the functionality, but everything should be in place if you do `] dev Variography` before running the notebook.

**Comment by juliohm** [comment]
@jheinen do you have an idea of how much work is missing to have polar coordinates support in GR? I am planning a workshop this year with GeoStats.jl and would like to avoid PyPlot.jl altogether. I would only ask attendees to install GR.jl in their Julia workstations.

**Comment by jheinen** [comment]
It should not be that much. Right now, I'm very busy, but in 2-3 weeks I can take a close look ...

**Comment by juliohm** [comment]
@jheinen any chance that you will have time available to work on this polar heatmap feature in the following month? The workshop will happen on the week of August 10, and that is the only plot type that is holding me from using only GR as the backend.

**Comment by jheinen** [comment]
GR will have support for polar heatmaps in a few days
<img width="612" alt="Screen Shot 2019-05-29 at 22 38 30" src="https://user-images.githubusercontent.com/5718244/58589872-31c6dc00-8263-11e9-9788-a6900ef4e3e5.png">
:

**Comment by juliohm** [comment]
Thank you @jheinen! This is awesome! I :heart: open-source.

**Comment by juliohm** [comment]
Should we close this issue in the GR.jl side? I understand that the only missing component is the connection with Plots.jl.

**Comment by jheinen** [comment]
@juliohm you mean [this](https://github.com/JuliaEarth/GeoStatsTutorials/blob/master/notebooks/DirectionalVariograms.ipynb) example?

**Comment by juliohm** [comment]
Yes, a modified version of this example that will include that last plot in polar coordinates.

**Comment by juliohm** [comment]
As you can see in that notebook, the PyPlot version stopped working. The polar heatmap that shows up is very tiny for some reason.

**Comment by jheinen** [comment]
How could I create 2 simple 2d array from `γp` which I could directly pass to GR (for testing with `GR.polarheatmap`)?

**Comment by juliohm** [comment]
If the object is

```julia
v = Varioplane(sdata, :property, maxlag=50.)
```

then the conversion is:

```julia
  # retrieve info
  θs = v.θs
  γs = v.γs

  # abscissa is the same for all variograms
  rs, _, __ = values(γs[1])

  # variogram values for all variograms
  zs = map(γs) do γ
    _, zs, __ = values(γ)

    # handle NaN values (i.e. empty bins)
    isnan(zs[1]) && (zs[1] = 0)
    for i in 2:length(zs)
      isnan(zs[i]) && (zs[i] = zs[i-1])
    end

    zs
  end
  Z = reduce(hcat, zs)

  # exploit symmetry
  θs = range(0, 2π, length=2*length(θs))
  Z  = [Z Z]

  θs, rs, Z
```

**Comment by jheinen** [comment]
Plots.GR.polarheatmap(Z) shows
<img width="712" alt="Screen Shot 2020-06-11 at 03 11 01" src="https://user-images.githubusercontent.com/5718244/84333914-51d8ed00-ab91-11ea-9048-3fddee9bd28f.png">
Is this what you would expect?

**Comment by juliohm** [comment]
Is there a way to pass the radius and angles? The issue here is that the matrix Z at the end is not square. It has a number of lines corresponding to the radii and a number of columns corresponding to the angles. If you pass only the Z matrix will the polarheatmap command know? Maybe transposing the matrix would solve? The expected result is in the top of this thread.

**Comment by juliohm** [comment]
Will try to play with it locally as well, didn't know we had access to the GR functions from within Plots.

**Comment by jheinen** [comment]
Ah, I see: `Plots.GR.polarheatmap(Z')` looks better:
<img width="712" alt="Screen Shot 2020-06-11 at 03 20 42" src="https://user-images.githubusercontent.com/5718244/84334396-95802680-ab92-11ea-8496-a820fad23a88.png">
So the problem is related to the color handling in the Plots GR backend. In principle, it should work.

**Comment by juliohm** [comment]
That is so much prettier than PyPlot :heart_eyes: Any chance this can get merged and released?

**Comment by juliohm** [comment]
Also I think the radii grid lines aren't showing the correct labels since we are not passing the vector of radii.

**Comment by jheinen** [comment]
I used the official GR version and bypassed Plots (`Plots.GR.polarheatmap()`). I can try to fix the bug within Plots, but a lot of things have changed in the `gr` backend code and probably someone other has to help.

**Comment by jheinen** [comment]
I must confess, I lost track of something in the Plots GR part ...

**Comment by juliohm** [comment]
I will open an issue in Plots.jl and point to this one here. I was also wondering... Now that the recipe system is decoupled from Plots.jl (as far as I understand), could GR.jl implement the recipes directly? Personally, I only use GR + recipes, so if the recipes existed in this repository I could just do

```julia
using GR
```

as opposed to

```julia
using Plots
gr()
```

Makes sense?

What we can do about the radii and angle vectors? Am I correct that the current `GR.polarheatmap` function does not accept these custom coordinates?

**Comment by juliohm** [comment]
Also, what can we do about the radii and angle vectors? Am I correct that the current `GR.polarheatmap` function does not accept custom coordinates?

**Comment by jheinen** [comment]
> Also I think the radii grid lines aren't showing the correct labels since we are not passing the vector of radii.

The current implementation only allows to specify the limits - θs and rs are assumed to be equally spaced.

**Comment by jheinen** [comment]
This is what the non-polar representation would look like:
```
X = LinRange(0, 2π, 101)
Y = LinRange(0, 50, 21)
Plots.GR.heatmap(X, Y, Z, xlim=(0,2pi))
```
<img width="712" alt="Screen Shot 2020-06-11 at 07 43 43" src="https://user-images.githubusercontent.com/5718244/84349309-52847a00-abb7-11ea-8a00-90b669947223.png">

`X` and `Y` could be non-equally spaced and would be properly handled by `heatmap`.

**Comment by juliohm** [comment]
> The current implementation only allows to specify the limits - θs and rs are assumed to be equally spaced.

Assuming equal spacing is perfectly fine. What I am wondering is if we could fix the labels of the radii grids depending on a vector of radii? Could we have a keyword argument to polarheatmap maybe that fixes the labels?

**Comment by juliohm** [comment]
@jheinen do you think it is possible to have this label option in polarheatmap? Or maybe I could hide the labels in the plot recipe meanwhile? Thank you for your help as usual.

**Comment by jheinen** [comment]
You mean 0..50 instead of 0..20.5, right?

**Comment by juliohm** [comment]
Exactly. The range of the radii.

**Comment by jheinen** [comment]
The quick and dirty solution is ```Plots.GR.polarheatmap(Z', ylim=(0, 50))```. I will see how we can use this within `Plots`.
<img width="712" alt="Screen Shot 2020-06-15 at 08 44 49" src="https://user-images.githubusercontent.com/5718244/84625938-7e1fa100-aee4-11ea-83f2-2e3c94528a9c.png">

**Comment by jheinen** [comment]
I tried to fix the problem in [this](https://github.com/JuliaPlots/Plots.jl/pull/2788) PR.
